### PR TITLE
Refactor: High Performance Order Storage

### PR DESCRIPTION
**Pull request by nurul-umbhiya**
### All Submissions:

* [ ] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [ ] My code satisfies feature requirements
* [ ] My code is tested
* [ ] My code passes the PHPCS tests
* [ ] My code has proper inline documentation
* [ ] I've included developer documentation (optional)
* [ ] I've added proper labels to this pull request

### Changes proposed in this Pull Request:

1. HPOS Rewrite For Dokan Lite
2. Refactor Existing Codebase
3. Commission Calculation based on order data

Related Dokan Pro PR: https://github.com/weDevsOfficial/dokan-pro/pull/2300

### Related Resource Link

1. https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book
2. https://github.com/woocommerce/woocommerce/wiki/HPOS:-new-order-querying-APIs
3. https://developer.woocommerce.com/2022/10/11/hpos-upgrade-faqs/
5. https://developer.woocommerce.com/2022/09/14/high-performance-order-storage-progress-report/
6. https://developer.woocommerce.com/2023/03/17/performance-benchmarking-for-woocommerce-hpos/

Closes https://github.com/weDevsOfficial/dokan-pro/issues/2242
Closes https://github.com/weDevsOfficial/dokan-pro/issues/1414
Closes https://github.com/weDevsOfficial/dokan-pro/issues/2378
Closes https://github.com/weDevsOfficial/dokan/issues/1948
Closes https://github.com/weDevsOfficial/client-issue/issues/2

### How to test the changes in this Pull Request:

Test -> Orders, Reports, Logs, Commission under Admin,
Test -> Order List Page, Order Details Page under vendor dashboard
Test -> Commission calculation 
Test -> Refund (Partial/Full)

Test the above feature without enabling the HPOS feature.
Test the above feature after enabling the HPOS feature

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

### CODE REVIEW

File: includes/wc-legacy-functions.php [hunk 1]
```diff
@@ -9,50 +9,12 @@
...
-    }
-
-    // flag it as it has a suborder
-    update_post_meta( $parent_order_id, 'has_sub_order', true );
-
-    // seems like we've got multiple sellers
-    foreach ( $sellers as $seller_id => $seller_products ) {
-        dokan_create_seller_order( $parent_order, $seller_id, $seller_products );
-    }
+    wc_deprecated_function( 'dokan_create_sub_order', '3.7.13', 'dokan()->order->create_sub_order()' );
```
Comment (shohag121) [inline includes/wc-legacy-functions.php:17]: Bhai we should add `dokan()->order->create_sub_order()` below `wc_deprecated_function` as well.

File: includes/Admin/functions.php [hunk 2]
```diff
@@ -149,17 +165,18 @@
...
  * @param string $year
  * @param string $start
  * @param string $end
  *
+ * @deprecated DOKAN_SINCE
+ *
  * @return array
  */
 function dokan_admin_report( $group_by = 'day', $year = '', $start = '', $end = '' ) {
+    wc_deprecated_function( 'dokan_admin_report', '3.7.111' );
```
[the remark this answers is not in the document]
Comment (nurul-umbhiya) [reply includes/Admin/functions.php:130]: actually, there is no replacement for this function. we are getting report data from REST API; these methods were never used. So deprecating these functions makes more sense. We don't have to deal with the legacy code in the future.

File: includes/Order/Manager.php [hunk 3]
```diff
@@ -89,72 +109,80 @@
...
         }
 
-        //swap dates if start date is after end date
-        $date_filter_applied = false;
         if ( $date_from && $date_to ) {
-            // fix start and end date
+            $date_from = $date_from->format( 'Y-m-d' );
+            $date_to   = $date_to->format( 'Y-m-d' );
+            //swap dates if start date is after end date
             if ( $date_from > $date_to ) {
```
[the remark this answers is not in the document]
Comment (nurul-umbhiya) [reply includes/Order/Manager.php:144]: performance wise current one is faster,

File: includes/wc-legacy-functions.php (hunk 1 above)
Comment (nurul-umbhiya) [reply includes/wc-legacy-functions.php:17]: Those are legacy functions. I don't think we need to keep supporting those functions anymore. we will be removing those functions later on.

File: includes/Order/MiscHooks.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            return $counts;
+        }
+
+        $vendor_id = dokan_get_current_user_id();
+        if ( empty( $vendor_id ) ) {
+            return $counts;
+        }
+
+        // Current order counts for the vendor.
+        $vendor_order_counts = dokan_count_orders( $vendor_id );
```
[the remark this answers is not in the document]
Comment (nurul-umbhiya) [reply includes/Order/MiscHooks.php:182]: support is already added on the function level. so this function is safe to use.

File: includes/Order/MiscHooks.php [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+     * @return object $counts
+     */
+    public function modify_vendor_order_counts( $counts ) {
+        global $pagenow;
+
+        if ( current_user_can( 'manage_woocommerce' ) ) {
+            return $counts;
+        }
+
+        if ( 'edit.php' !== $pagenow || 'shop_order' !== get_query_var( 'post_type' ) ) {
```
[the remark this answers is not in the document]
Comment (nurul-umbhiya) [reply includes/Order/MiscHooks.php:172]: it was left alone intentionally. this feature will work only if hpos feature is disabled. to change order count for hpos, another hook was used.

**Review by shohag121 — APPROVED** [review]

**Pull request merged**